recipes/spirv-tools/all/conanfile.py

```diff
@@ -0,0 +1,122 @@
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "SPIRV-Tools-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def configure(self):
+        if self.settings.os == 'Windows':
+            del self.options.fPIC
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+
+        cmake = CMake(self)
+
+        if self.options.shared and self.settings.compiler == "Visual Studio":
+            cmake.definitions["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
+
+        # Required by the project's CMakeLists.txt
+        cmake.definitions["SPIRV-Headers_SOURCE_DIR"] = self.deps_cpp_info["spirv-headers"].rootpath
+
+        # There are some switch( ) statements that are causing errors
+        # need to turn this off
+        cmake.definitions["SPIRV_WERROR"] = False
+
+        cmake.definitions["SKIP_SPIRV_TOOLS_INSTALL"] = False
+        cmake.definitions["SPIRV_LOG_DEBUG"] = False
+        cmake.definitions["SPIRV_SKIP_TESTS"] = True
+        cmake.definitions["SPIRV_CHECK_CONTEXT"] = False
+        cmake.definitions["SPIRV_BUILD_FUZZER"] = False
+
+        cmake.configure(build_folder=self._build_subfolder)
+        self._cmake = cmake
+        return self._cmake
+
+    def build(self):
+        self._patch_sources()
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def _patch_sources(self):
+        # CMAKE_POSITION_INDEPENDENT_CODE was set ON for the entire
+        # project in the lists file.
+        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                              "set(CMAKE_POSITION_INDEPENDENT_CODE ON)", "")
+
+    def package(self):
+        self.copy(pattern="LICENSE*", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        # Error KB-H020, complaining that .pc files are found
+        tools.rmdir(os.path.join(self.package_folder, "lib/pkgconfig"))
+
+        # Error KB-H019, complaining that .pc files are found
+        tools.rmdir(os.path.join(self.package_folder, "lib/cmake"))
+
+        # SPIRV-Tools-shared is meant to be a shared-only c-only API for the library .
+        # it is built by the original CMakeLists.txt file. It is the same
+        # as the other libraries except only the C interface is exposed
+        # This library is being removed because it is only causing confusion.
+        # As a result, you are restricted to to using the C++ api only.
+        for bin_file in glob.glob(os.path.join(self.package_folder, "bin", "*SPIRV-Tools-shared.dll")):
+            os.remove(bin_file)
+        for lib_file in glob.glob(os.path.join(self.package_folder, "lib", "*SPIRV-Tools-shared*")):
+            os.remove(lib_file)
+
+    def package_info(self):
```

Comment:
```suggestion
    def package_info(self):
        # TODO: set targets names when components available in conan
        self.cpp_info.names["cmake_find_package"] = "SPIRV-Tools"
        self.cpp_info.names["cmake_find_package_multi"] = "SPIRV-Tools"
```

Vulkan-ValidationLayers excepts this target (which is a component actually). At least add TODO comment ;)

Reply:
Added and pushed

Reply:
Components is available

Reply:
I'm not sure what needs to be done for the components, @SpaceIm , could you make a suggestion?

Reply:
It's pretty new. From what I understand from https://blog.conan.io/2020/05/07/New-conan-release-1-25.html, I would try this:

```python

    def package_info(self):
        self.cpp_info.name = "SPIRV-Tools"
        self.cpp_info.components["spirv-tools"].name = "SPIRV-Tools"
        self.cpp_info.components["spirv-tools"].libs = ["SPIRV-Tools"]
        if self.settings.os == "Linux":
            self.cpp_info.components["spirv-tools"].system_libs.append("rt")
        if not self.options.shared and self._stdcpp_library:
            self.cpp_info.components["spirv-tools"].system_libs.append(self._stdcpp_library)
        self.cpp_info.components["spirv-tools-opt"].name = "SPIRV-Tools-opt"
        self.cpp_info.components["spirv-tools-opt"].libs = ["SPIRV-Tools-opt"]
        self.cpp_info.components["spirv-tools-opt"].requires = ["spirv-tools"]
        self.cpp_info.components["spirv-tools-link"].name = "SPIRV-Tools-link"
        self.cpp_info.components["spirv-tools-link"].libs = ["SPIRV-Tools-link"]
        self.cpp_info.components["spirv-tools-link"].requires = ["spirv-tools-opt"]
        self.cpp_info.components["spirv-tools-reduce"].name = "SPIRV-Tools-reduce"
        self.cpp_info.components["spirv-tools-reduce"].libs = ["SPIRV-Tools-reduce"]
        self.cpp_info.components["spirv-tools-reduce"].requires = ["spirv-tools", "spirv-tools-opt"]
        self.env_info.PATH.append(os.path.join(self.package_folder, "bin"))

    @property
    def _stdcpp_library(self):
        libcxx = self.settings.get_safe("compiler.libcxx")
        if libcxx in ("libstdc++", "libstdc++11"):
            return "stdc++"
        elif libcxx in ("libc++",):
            return "c++"
        else:
            return False
```

(`_stdcpp_library` is here to propagate c++ lib into `SPIRV-Tools` for the C API)

Reply:
Therefore, for the moment, I would do this:

```python
    def package_info(self):
        # TODO: set targets names when components available in conan
        self.cpp_info.names["cmake_find_package"] = "SPIRV-Tools"
        self.cpp_info.names["cmake_find_package_multi"] = "SPIRV-Tools"
        self.cpp_info.names["pkg_config"] = "SPIRV-Tools"
        self.cpp_info.libs = ["SPIRV-Tools-reduce", "SPIRV-Tools-link", "SPIRV-Tools-opt", "SPIRV-Tools"]
        if self.settings.os == "Linux":
            self.cpp_info.system_libs.append("rt") # for SPIRV-Tools
        if not self.options.shared and self._stdcpp_library:
            self.cpp_info.system_libs.append(self._stdcpp_library) # for SPIRV-Tools C API

        bin_path = os.path.join(self.package_folder, "bin")
        self.output.info("Appending PATH environment variable: {}".format(bin_path))
        self.env_info.path.append(bin_path)

    @property
    def _stdcpp_library(self):
        libcxx = self.settings.get_safe("compiler.libcxx")
        if libcxx in ("libstdc++", "libstdc++11"):
            return "stdc++"
        elif libcxx in ("libc++",):
            return "c++"
        else:
            return False
```

and in CMakeLists of test_package, you can remove `set_property(TARGET ${PROJECT_NAME}_c PROPERTY LINKER_LANGUAGE CXX)`

Reply:
By the way, if components are used in this recipe, maybe we should check conan version. Because I tried first with 1.24.0, I was able to create the package without any error or warning, but libs ordering was incorrect with cmake generator.

Reply:
I'm wondering if we should use components because the feature is still listed as experimental..

Reply:
I agree.
@uilianries What do you think?

Reply:
No problem including it. but global info is not allowed, you will need to remove:
```
self.cpp_info.names["cmake_find_package"] = "SPIRV-Tools"       # not sure, SPIRV-Tools doesn't
        self.cpp_info.names["cmake_find_package_multi"] = "SPIRV-Tools" # export a namespace
        self.cpp_info.names["pkg_config"] = "SPIRV-Tools"
```

Reply:
global names are allowed I think (at least it properly work with conan 1.25.0), how would you define a namespace without that?